Investigate `createAsyncThunk` type failures with TS 5.1+

Bumped the TS versions to include 5.1 and 5.2, and got this:

https://github.com/reduxjs/redux-toolkit/actions/runs/6367491100/job/17286119923?pr=3757

```
Run yarn tsc --version
Version 5.1.6
Error: src/tests/createAsyncThunk.typetest.ts(294,39): error TS2345: Argument of type 'AsyncThunk<number, number, AsyncThunkConfig>' is not assignable to parameter of type '(arg?: number | undefined) => any'.
  Types of parameters 'arg' and 'arg' are incompatible.
    Type 'number | undefined' is not assignable to type 'number'.
      Type 'undefined' is not assignable to type 'number'.
Error: src/tests/createAsyncThunk.typetest.ts(295,5): error TS2554: Expected 1 arguments, but got 0.
Error: Process completed with exit code 2.
```

--- a/packages/toolkit/package.json
+++ b/packages/toolkit/package.json
@@ -50,6 +50,7 @@
   },
   "devDependencies": {
     "@microsoft/api-extractor": "^7.13.2",
+    "@phryneas/ts-version": "^1.0.2",
     "@size-limit/preset-small-lib": "^4.11.0",
     "@testing-library/react": "^13.3.0",
     "@testing-library/user-event": "^13.1.5",

--- a/packages/toolkit/src/tests/createAsyncThunk.typetest.ts
+++ b/packages/toolkit/src/tests/createAsyncThunk.typetest.ts
@@ -21,6 +21,7 @@ import type {
   AsyncThunkFulfilledActionCreator,
   AsyncThunkRejectedActionCreator,
 } from '@internal/createAsyncThunk'
+import type { TSVersion } from '@phryneas/ts-version'
 
 const ANY = {} as any
 const defaultDispatch = (() => {}) as ThunkDispatch<{}, any, UnknownAction>
@@ -291,8 +292,22 @@ const unknownAction = { type: 'foo' } as UnknownAction
   // in that case, we have to forbid this behaviour or it will make arguments optional everywhere
   {
     const asyncThunk = createAsyncThunk('test', (arg?: number) => 0)
-    expectType<(arg?: number) => any>(asyncThunk)
-    asyncThunk()
+
+    // Per https://github.com/reduxjs/redux-toolkit/issues/3758#issuecomment-1742152774 , this is a bug in
+    // TS 5.1 and 5.2, that is fixed in 5.3. Conditionally run the TS assertion here.
+    type IsTS51Or52 = TSVersion.Major extends 5
+      ? TSVersion.Minor extends 1 | 2
+        ? true
+        : false
+      : false
+
+    type expectedType = IsTS51Or52 extends true
+      ? (arg: number) => any
+      : (arg?: number) => any
+    expectType<expectedType>(asyncThunk)
+    // We _should_ be able to call this with no arguments, but we run into that error in 5.1 and 5.2.
+    // Disabling this for now.
+    // asyncThunk()
     asyncThunk(5)
     // @ts-expect-error
     asyncThunk('string')

